File: javascript/28.find-the-index-of-the-first-occurrence-in-a-string.js
/*
 * @lc app=leetcode id=28 lang=javascript
 *
 * [28] Find the Index of the First Occurrence in a String
 */

// @lc code=start
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function (haystack, needle) {
  // Solution 1:
  // const occur = [];
  // for (let i = 0; i < haystack.length; i++) {
  //   if (haystack[i] === needle[0]) {
  //     occur.push(i);
  //   }
  // }
  // for (let i of occur) {
  //   if (i + needle.length - 1 > haystack.length) break;
  //   for (let j = 0; j < needle.length; j++) {
  //     if (haystack[i + j] !== needle[j]) break;
  //     if (j === needle.length - 1) return i;
  //   }
  // }
  // return -1;

  // Solution 2:
  return haystack.indexOf(needle);
};
// @lc code=end

File: rust/28.find-the-index-of-the-first-occurrence-in-a-string.rs
/*
 * @lc app=leetcode id=28 lang=rust
 *
 * [28] Find the Index of the First Occurrence in a String
 */

// @lc code=start
impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        match haystack.find(&needle) {
            Some(i) => return i as i32,
            None => return -1,
        }
    }
}
// @lc code=end
